### ISSUE TITLE
Invalid argument: Expects arg[1] to be bool but float is provided

### ISSUE THREAD

**Issue by tangjie77wd** [tensorflow/tensorflow#30273]
**Tensorflow version:1.12 .I want to translate my python code to c++ version and here is my python code below:**
`
import tensorflow as tf
import  numpy as np
from skimage import io, transform
import cv2
#import infer
sess = tf.Session()
new_saver = tf.train.import_meta_graph(r'D:\CNN_test_result_sizhuangzao_model\v2_101_ziji_9\model\model.meta')
new_saver.restore(sess,tf.train.latest_checkpoint(r'D:\CNN_test_result_sizhuangzao_model\v2_101_ziji_9\model'))

graph = tf.get_default_graph()
input_x = graph.get_tensor_by_name("input:0")
is_training_x=graph.get_tensor_by_name("is_training:0")
#print (sess.run("input:0"))

jpg_path=r"D:\CNN_test_result_sizhuangzao\chaodacesiji1\pos.txt"
file = open(jpg_path)
lines = file.readlines()
num=0
for line in lines: 
    img0 = io.imread(line.strip())
    l=img0.shape
    k=l[0]
    c=l[1]
    num+=1
    print(num)
    if  c/k<5:
        continue
    img = transform.resize(img0, (48,160, 1))
    feed_dict={input_x:np.reshape(img, [-1, 48, 160, 1]),is_training_x:False}

    prob_op = graph.get_operation_by_name('output')
    out_softmax = graph.get_tensor_by_name("output:0")
    img_out_softmax = sess.run(out_softmax,feed_dict)
    prediction_labels = np.argmax(img_out_softmax,1)
`

**Now this is my c++ version below (There is some wrong but i do not know where it is !)**

`
#include "tensorflow/core/framework/graph.pb.h"
#include <tensorflow/core/public/session_options.h>
#include <tensorflow/core/protobuf/meta_graph.pb.h>
#include <fstream>
#include <utility>
#include <vector>
#include <Eigen/Core>
#include <Eigen/Dense>

#include "tensorflow/cc/ops/const_op.h"
#include "tensorflow/cc/ops/image_ops.h"
#include "tensorflow/cc/ops/standard_ops.h"
#include "tensorflow/core/framework/graph.pb.h"
#include "tensorflow/core/framework/tensor.h"
#include "tensorflow/core/graph/default_device.h"
#include "tensorflow/core/graph/graph_def_builder.h"
#include "tensorflow/core/lib/core/errors.h"
#include "tensorflow/core/lib/core/stringpiece.h"
#include "tensorflow/core/lib/core/threadpool.h"
#include "tensorflow/core/lib/io/path.h"
#include "tensorflow/core/lib/strings/stringprintf.h"
#include "tensorflow/core/platform/env.h"
#include "tensorflow/core/platform/init_main.h"
#include "tensorflow/core/platform/logging.h"
#include "tensorflow/core/platform/types.h"
#include "tensorflow/core/public/session.h"
#include "tensorflow/core/util/command_line_flags.h"

using namespace std;
using namespace tensorflow;
using namespace tensorflow::ops;
using tensorflow::Flag;
using tensorflow::Tensor;
using tensorflow::Status;
using tensorflow::string;
using tensorflow::int32;

typedef std::vector<std::pair<std::string, tensorflow::Tensor>> tensor_dict;
using tensorflow::Status;

static Status ReadtheFile(tensorflow::Env* env, const string& filename,Tensor* output) {
  tensorflow::uint64 file_size = 0;
  TF_RETURN_IF_ERROR(env->GetFileSize(filename, &file_size));

  string contents;
  contents.resize(file_size);

  std::unique_ptr<tensorflow::RandomAccessFile> file;
  TF_RETURN_IF_ERROR(env->NewRandomAccessFile(filename, &file));

  tensorflow::StringPiece data;
  TF_RETURN_IF_ERROR(file->Read(0, file_size, &data, &(contents)[0]));
  if (data.size() != file_size) {
    return tensorflow::errors::DataLoss("Truncated read of '", filename,
                                        "' expected ", file_size, " got ",
                                        data.size());
  }
//  output->scalar<string>()() = data.ToString();
  output->scalar<string>()() = string(data);
  return Status::OK();
}

Status ReadImageFile(const string& file_name, const int input_height,
                               const int input_width, const float input_mean,
                               const float input_std,
                               std::vector<Tensor>* out_tensors) {
  auto root = tensorflow::Scope::NewRootScope();
  using namespace ::tensorflow::ops;

  string input_name = "file_reader";
  string output_name = "normalized";

  // read file_name into a tensor named input
  Tensor input(tensorflow::DT_STRING, tensorflow::TensorShape());
  TF_RETURN_IF_ERROR(ReadtheFile(tensorflow::Env::Default(), file_name, &input));

  // use a placeholder to read input data
  auto file_reader =Placeholder(root.WithOpName("input"), tensorflow::DataType::DT_STRING);

  std::vector<std::pair<string, tensorflow::Tensor>> inputs = {{"input", input},};

  // Now try to figure out what kind of file it is and decode it.
    const int wanted_channels = 1;
    tensorflow::Output image_reader;
	if (tensorflow::str_util::EndsWith(file_name, ".png"))
	{
	  image_reader = DecodePng(root.WithOpName("png_reader"), file_reader,
							   DecodePng::Channels(wanted_channels));
	}
	else if (tensorflow::str_util::EndsWith(file_name, ".gif"))
	{
	  // gif decoder returns 4-D tensor, remove the first dim
	  image_reader =
		  Squeeze(root.WithOpName("squeeze_first_dim"),
				  DecodeGif(root.WithOpName("gif_reader"), file_reader));
	}
	else if (tensorflow::str_util::EndsWith(file_name, ".bmp"))
	{
	  image_reader = DecodeBmp(root.WithOpName("bmp_reader"), file_reader);
	}
	else
	{
	  // Assume if it's neither a PNG nor a GIF then it must be a JPEG.
	  image_reader = DecodeJpeg(root.WithOpName("jpeg_reader"), file_reader,
								DecodeJpeg::Channels(wanted_channels));
	}
  // Now cast the image data to float so we can do normal math on it.
  auto float_caster =Cast(root.WithOpName("float_caster"), image_reader, tensorflow::DT_FLOAT);

  auto dims_expander = ExpandDims(root.WithOpName("expand"), float_caster, 0);

  float input_max = 255;
  Div(root.WithOpName("div"),dims_expander,input_max);

  tensorflow::GraphDef graph;
  TF_RETURN_IF_ERROR(root.ToGraphDef(&graph));

  std::unique_ptr<tensorflow::Session> session(
      tensorflow::NewSession(tensorflow::SessionOptions()));
  TF_RETURN_IF_ERROR(session->Create(graph));
//  std::vector<Tensor> out_tensors;
//  TF_RETURN_IF_ERROR(session->Run({}, {output_name + ":0", output_name + ":1"},
//                                    {}, &out_tensors));
  TF_RETURN_IF_ERROR(session->Run({inputs}, {"div"}, {}, out_tensors));
  return Status::OK();
}

int main()
{
  Session* session;
  Status status = NewSession(SessionOptions(), &session);

  const std::string graph_fn = "/media/root/Ubuntu311/projects/Ecology_projects/tensorflowtest/model-0617/model.meta";
  MetaGraphDef graphdef;
  Status status_load = ReadBinaryProto(Env::Default(), graph_fn, &graphdef); //从meta文件中读取图模型;
  if (!status_load.ok()) {
        std::cout << "ERROR: Loading model failed..." << graph_fn << std::endl;
        std::cout << status_load.ToString() << "\n";
        return -1;
  }

  Status status_create = session->Create(graphdef.graph_def()); //将模型导入会话Session中;
  if (!status_create.ok()) {
        std::cout << "ERROR: Creating graph in session failed..." << status_create.ToString() << std::endl;
        return -1;
  }
  cout << "Session successfully created.Load model successfully!"<< endl;

  // 读入预先训练好的模型的权重
  const std::string checkpointPath = "/media/root/Ubuntu311/projects/Ecology_projects/tensorflowtest/model-0617/model";
  Tensor checkpointPathTensor(DT_STRING, TensorShape());
  checkpointPathTensor.scalar<std::string>()() = checkpointPath;
  status = session->Run(
		  {{ graphdef.saver_def().filename_tensor_name(), checkpointPathTensor },},
		  {},{graphdef.saver_def().restore_op_name()},nullptr);
  if (!status.ok())
  {
	  throw runtime_error("Error loading checkpoint from " + checkpointPath + ": " + status.ToString());
  }
  cout << "Load weights successfully!"<< endl;

  //read image for prediction...
  string image_path= "/media/root/Ubuntu311/projects/Ecology_projects/copy/cnn-imgs/AABW.jpg";
   int input_height =48;
   int input_width=160;
   int input_mean=0;
   int input_std=1;
   std::vector<Tensor> resized_tensors;
   Status read_tensor_status =
       ReadImageFile(image_path, input_height, input_width, input_mean,
                               input_std, &resized_tensors);
   if (!read_tensor_status.ok()) {
     LOG(ERROR) << read_tensor_status;
     cout<<"resing error"<<endl;
     return -1;
   }

   const Tensor& resized_tensor = resized_tensors[0];
   std::cout <<"Read image successfully: "<< resized_tensor.DebugString()<<endl;

   std::string Input1Name = "input";
   std::string Input2Name = "is_training";
   vector<std::pair<string, Tensor> > inputs;
   inputs.push_back(std::make_pair(Input1Name, resized_tensor));
   inputs.push_back(std::make_pair(Input2Name, resized_tensor));

   vector<tensorflow::Tensor> outputs;
   string output2="out_softmax";
   string output_ = "output";
   Status status_run = session->Run(inputs, {output2}, {}, &outputs);
   if (!status_run.ok()) {
       std::cout << "ERROR: RUN failed..."  << std::endl;
       std::cout << status_run.ToString() << "\n";
       return -1;
   }
   //Fetch output value
   std::cout << "Output tensor size:" << outputs.size() << std::endl;
   for (std::size_t i = 0; i < outputs.size(); i++) {
       std::cout <<"result: "<<i<<" :"<< outputs[i].DebugString()<<endl;
   }
  cout << "Prediction successfully!"<< endl;

  return 0;
}
`

I built the c++ version successfully but  the error informations came to me----- " ERROR: RUN failed...
Invalid argument: Expects arg[1] to be bool but float is provided ". Any help will be much appreciated.

**Comment by rmothukuru** [comment]
This question is better asked on [StackOverflow](http://stackoverflow.com/questions/tagged//tensorflow) since it is not a bug or feature request. There is also a larger community that reads questions there.

**Comment by tangjie77wd** [comment]
Thanks,i have put it here [https://stackoverflow.com/questions/56880358/model-pruner-failed-invalid-argument-invalid-input-graph-when-run-tensorflow-c](url)

**Comment by rmothukuru** [comment]
@tangjie77wd ,
Thank you. Closing the issue as it is a support question and as you have raised it in Stack Overflow.

**Comment by Mr-DDDAlKilanny** [comment]
Maybe the problem is in your input / output names are not correct

**Comment by tangjie77wd** [comment]
Has been solved!

**Comment by felix288** [comment]
@tangjie77wd Hi, how did you actually solve this issue? I have encountered a similar issue.  Many thanks

**Comment by panshaohua** [comment]
@tangjie77wd  Could you tell how to slove this issue?

**Comment by panshaohua** [comment]
@felix288 Could you solve you problem?

**Comment by tangjie77wd** [comment]
you can link here :https://blog.csdn.net/wd1603926823/article/details/99406974 or&nbsp;https://blog.csdn.net/wd1603926823/article/details/98086550




------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"panshaohua"<notifications@github.com&gt;;
发送时间:&nbsp;2020年5月12日(星期二) 晚上9:52
收件人:&nbsp;"tensorflow/tensorflow"<tensorflow@noreply.github.com&gt;;
抄送:&nbsp;"巴卫"<1603926823@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [tensorflow/tensorflow] Invalid argument: Expects arg[1] to be bool but float is provided (#30273)





 
@tangjie77wd  Could you tell how to slove this issue?
